posts/2022-10-04-microprofile-serverless-ibm-code-engine.adoc

```diff
@@ -0,0 +1,412 @@
+
+```
+
+With our data in string format, we can now get the `byte[]` representation of the data.
+
+The `CloudEventBuilder` class provides the necessary components to build our CloudEvent.  Use the most recent specification version, which is `v1()`.
+```
+    return CloudEventBuilder.v1()
+        .withData(jsonString.getBytes())
+        .withDataContentType("application/json")
+        .withId("properties")
+        .withType("java.properties")
+        .withSource(URI.create("http://system.poperties"))
+        .build();
+```
+Besides `withData()`, the rest of the methods set the values that will be returned as headers in the response.
+Once all the required properties are set, you can build the CloudEvent Object.
+
+If you are missing any required properties when your application attempts to build the CloudEvent, an exception will be thrown that informs you of a missing property. To see which properties are required, you can review the https://github.com/cloudevents/spec/blob/v1.0/spec.md[specification].
+
+Because CloudEvents can be from a wide variety of sources that might differ even within a single provider, the majority of the fields are fairly free-form.
+
+==== Receiving a CloudEvent
+Now that we've returned a CloudEvent, how is it handled within the application?
+
+First, we'll enhance the `SystemResource` class to add a query method that can send a POST request with a body that contains the system properties that we want returned.
+
+The body of the request will be a JSON array that contains each property we want as an a member.
+```
+["java.vendor.url","awt.toolkit"]
+```
+
+Add the following method declaration to the `SystemResource` class.
+```
+    @POST
+    @Produces(MediaType.APPLICATION_JSON)
+    @Timed(name = "queryPropertiesTime",
+            description = "Time needed to query the JVM system properties")
+    @Counted(absolute = true, description
+            = "Number of times the JVM system properties are queried")
+    public CloudEvent queryProperties(CloudEvent query){
+
+
+}
+```
+In this case, we will return a CloudEvent, but you can return any type that confirms the request was received, such as `Response.ok().build();`
+
+Inside the method, we need to do some of the same things that we did within  the `getProperties()` method. But we also must handle the CloudEvent input.
+
+For some initial structure, add this block to the top of the method.
+```
+Map properties = System.getProperties();
+HashMap<String,String> props = new HashMap<>((Map<String,String>)properties);
+HashMap<String,String> qProps = new HashMap<String,String>();
+Jsonb jsonb = JsonbBuilder.create();
+```
+This gives us the Map of the properties, but in a form that is more useful later as we need to do more processing then we did within `getProperties()`
+
+To retrieve the data from the CloudEvent, we use `.withData();` to extra and Object of type CloudEventData, the data portion is in binary format, so needs to be converted to make it usable.
```

Comment:
^^ something seems off with this sentence but I'm not sure of the intended meaning